Applications/Spire/Source/KeyBindings/TaskKeysTableView.cpp

```diff
@@ -223,6 +224,125 @@ namespace {
         m_region(std::move(region)) {}
   };
 
+  class BreakoutInputBox : public QWidget {
+    public:
+      explicit BreakoutInputBox(AnyInputBox& input_box,
+          QWidget* parent = nullptr)
+          : QWidget(parent),
+            m_input_box(&input_box),
+            m_is_setting_read_only(false) {
+        m_breakout_box = new BreakoutBox(*m_input_box, this);
+        enclose(*this, *m_breakout_box);
+        setFocusProxy(m_breakout_box);
+        proxy_style(*this, *m_breakout_box);
+      }
+
+      const std::shared_ptr<AnyValueModel>& get_current() const {
+        return m_input_box->get_current();
+      }
+
+      const AnyRef& get_submission() const {
+        return m_input_box->get_submission();
+      }
+
+      void set_placeholder(const QString& placeholder) {
+        m_input_box->set_placeholder(placeholder);
+      }
+
+      bool is_read_only() const {
+        return m_input_box->is_read_only();
+      }
+
+      void set_read_only(bool read_only) {
+        if(m_is_setting_read_only) {
+          return;
+        }
+        m_is_setting_read_only = true;
+        if(read_only == is_read_only()) {
+          return;
+        }
+        m_input_box->set_read_only(read_only);
+        if(read_only) {
+          m_input_box->setFixedHeight(
+            m_input_box->minimumSizeHint().height());
+          restore();
+        } else {
+          m_input_box->setMinimumHeight(0);
+          m_input_box->setMaximumHeight(QWIDGETSIZE_MAX);
+          breakout();
+        }
+        m_is_setting_read_only = false;
+      }
+
+      connection connect_submit_signal(
+          const AnyInputBox::SubmitSignal::slot_type& slot) const {
+        return m_input_box->connect_submit_signal(slot);
+      }
+
+      connection connect_reject_signal(
+          const AnyInputBox::RejectSignal::slot_type& slot) const {
+        return m_input_box->connect_reject_signal(slot);
+      }
+
+    protected:
+      bool eventFilter(QObject* watched, QEvent* event) {
+        if(watched == m_input_box && event->type() == QEvent::KeyPress) {
+          auto& key_event = *static_cast<QKeyEvent*>(event);
+          if(key_event.key() == Qt::Key_Escape) {
+            if(auto parent = parentWidget()) {
+              QCoreApplication::sendEvent(parent, event);
+            }
+          }
+        } else if(watched == m_breakout_box &&
+            m_breakout_box->is_broken_out()) {
+          if(event->type() == QEvent::FocusIn) {
+            m_input_box->setFocus();
+          } else if(event->type() == QEvent::KeyPress) {
+            if(auto focus_proxy = find_focus_proxy(*m_input_box)) {
+              QCoreApplication::sendEvent(focus_proxy, event);
+              return true;
+            }
+          }
+        }
+        return QWidget::eventFilter(watched, event);
+      }
+
+    private:
+      BreakoutBox* m_breakout_box;
+      AnyInputBox* m_input_box;
```

Comment:
Refrain from duplicating state. Instead access the input box through the breakout box.